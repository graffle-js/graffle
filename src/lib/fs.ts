import fs from 'node:fs/promises'
import { extname, isAbsolute, join } from 'node:path'
import { errorFromMaybeError } from './prelude.js'

export const statMaybeExists = async (path: string) => {
  return await fs.stat(path).catch((_: unknown) => {
    const error = errorFromMaybeError(_)
    return `code` in error && typeof error.code === `string` && error.code === `ENOENT`
      ? null
      : Promise.reject(error)
  })
}

export const fileExists = async (path: string) => {
  return Boolean(await statMaybeExists(path))
}

export const isPathToADirectory = async (path: string) => {
  const stat = await fs.stat(path)
  return stat.isDirectory()
}

export const toAbsolutePath = (cwd: string, maybeAbsolutePath: string) =>
  isAbsolute(maybeAbsolutePath) ? maybeAbsolutePath : join(cwd, maybeAbsolutePath)

const isFileLikePath = (path: string) => {
  return Boolean(extname(path))
}

export const toFilePath = (fileName: string, path: string) => {
  if (isFileLikePath(path)) {
    return path
  } else {
    return join(path, fileName)
  }
}

// export const toFileChecked = async (fileName: string, path: string) => {
//   if (isFileLikePath(path)) {
//     return path
//   } else {
//     return join(path, fileName)
//   }
// }
