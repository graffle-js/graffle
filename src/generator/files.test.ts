import { readFile } from 'fs/promises'
import { expect, test } from 'vitest'
import { generateFiles } from './files.js'

test(`generates types from GraphQL SDL file`, async () => {
  await generateFiles({
    sourceDirPath: `./tests/ts/_/schema`,
    outputDirPath: `./tests/ts/_/schema/generated`,
    code: {
      libraryPaths: {
        schema: `../../../../../src/Schema/__.js`,
        scalars: `../../../../../src/Schema/NamedType/Scalar/Scalar.js`,
      },
    },
  })
  expect(
    await readFile(`./tests/ts/_/schema/generated/SchemaBuildtime.ts`, `utf8`),
  ).toMatchSnapshot()
  expect(
    await readFile(`./tests/ts/_/schema/generated/Scalar.ts`, `utf8`),
  ).toMatchSnapshot()
  expect(
    await readFile(`./tests/ts/_/schema/generated/SchemaRuntime.ts`, `utf8`),
  ).toMatchSnapshot()
})
