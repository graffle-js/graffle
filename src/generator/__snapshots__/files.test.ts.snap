// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates types from GraphQL SDL file 1`] = `
"import type * as $ from '../../../../../src/Schema/__.js'
import type * as $Scalar from './Scalar.ts'

// ------------------------------------------------------------ //
//                             Root                             //
// ------------------------------------------------------------ //

export namespace Root {
  export type Query = $.Object$2<'Query', {
    date: $.Field<$.Output.Nullable<$Scalar.Date>, null>
    dateNonNull: $.Field<$Scalar.Date, null>
    dateList: $.Field<$.Output.Nullable<$.Output.List<$.Output.Nullable<$Scalar.Date>>>, null>
    dateObject1: $.Field<$.Output.Nullable<Object.DateObject1>, null>
    dateUnion: $.Field<$.Output.Nullable<Union.DateUnion>, null>
    dateInterface1: $.Field<$.Output.Nullable<Interface.DateInterface1>, null>
    dateListNonNull: $.Field<$.Output.List<$Scalar.Date>, null>
    dateArg: $.Field<
      $.Output.Nullable<$Scalar.Date>,
      $.Args<{
        date: $.Input.Nullable<$Scalar.Date>
      }>
    >
    dateArgNonNull: $.Field<
      $.Output.Nullable<$Scalar.Date>,
      $.Args<{
        date: $Scalar.Date
      }>
    >
    dateArgList: $.Field<
      $.Output.Nullable<$Scalar.Date>,
      $.Args<{
        date: $.Input.Nullable<$.Input.List<$.Input.Nullable<$Scalar.Date>>>
      }>
    >
    dateArgNonNullList: $.Field<
      $.Output.Nullable<$Scalar.Date>,
      $.Args<{
        date: $.Input.List<$.Input.Nullable<$Scalar.Date>>
      }>
    >
    dateArgNonNullListNonNull: $.Field<
      $.Output.Nullable<$Scalar.Date>,
      $.Args<{
        date: $.Input.List<$Scalar.Date>
      }>
    >
    dateArgInputObject: $.Field<
      $.Output.Nullable<$Scalar.Date>,
      $.Args<{
        input: $.Input.Nullable<InputObject.InputObject>
      }>
    >
    InputObjectNested: $.Field<
      $.Output.Nullable<$Scalar.ID>,
      $.Args<{
        input: $.Input.Nullable<InputObject.InputObjectNested>
      }>
    >
    InputObjectNestedNonNull: $.Field<
      $.Output.Nullable<$Scalar.ID>,
      $.Args<{
        input: InputObject.InputObjectNested
      }>
    >
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
    idNonNull: $.Field<$Scalar.ID, null>
    string: $.Field<$.Output.Nullable<$Scalar.String>, null>
    stringWithRequiredArg: $.Field<
      $.Output.Nullable<$Scalar.String>,
      $.Args<{
        string: $Scalar.String
      }>
    >
    stringWithArgs: $.Field<
      $.Output.Nullable<$Scalar.String>,
      $.Args<{
        string: $.Input.Nullable<$Scalar.String>
        int: $.Input.Nullable<$Scalar.Int>
        float: $.Input.Nullable<$Scalar.Float>
        boolean: $.Input.Nullable<$Scalar.Boolean>
        id: $.Input.Nullable<$Scalar.ID>
      }>
    >
    stringWithArgEnum: $.Field<
      $.Output.Nullable<$Scalar.String>,
      $.Args<{
        ABCEnum: $.Input.Nullable<Enum.ABCEnum>
      }>
    >
    stringWithListArg: $.Field<
      $.Output.Nullable<$Scalar.String>,
      $.Args<{
        ints: $.Input.Nullable<$.Input.List<$.Input.Nullable<$Scalar.Int>>>
      }>
    >
    stringWithListArgRequired: $.Field<
      $.Output.Nullable<$Scalar.String>,
      $.Args<{
        ints: $.Input.List<$.Input.Nullable<$Scalar.Int>>
      }>
    >
    stringWithArgInputObject: $.Field<
      $.Output.Nullable<$Scalar.String>,
      $.Args<{
        input: $.Input.Nullable<InputObject.InputObject>
      }>
    >
    stringWithArgInputObjectRequired: $.Field<
      $.Output.Nullable<$Scalar.String>,
      $.Args<{
        input: InputObject.InputObject
      }>
    >
    listListIntNonNull: $.Field<$.Output.List<$.Output.List<$Scalar.Int>>, null>
    listListInt: $.Field<
      $.Output.Nullable<$.Output.List<$.Output.Nullable<$.Output.List<$.Output.Nullable<$Scalar.Int>>>>>,
      null
    >
    listInt: $.Field<$.Output.Nullable<$.Output.List<$.Output.Nullable<$Scalar.Int>>>, null>
    listIntNonNull: $.Field<$.Output.List<$Scalar.Int>, null>
    object: $.Field<$.Output.Nullable<Object.Object1>, null>
    objectNonNull: $.Field<Object.Object1, null>
    objectNested: $.Field<$.Output.Nullable<Object.ObjectNested>, null>
    objectWithArgs: $.Field<
      $.Output.Nullable<Object.Object1>,
      $.Args<{
        string: $.Input.Nullable<$Scalar.String>
        int: $.Input.Nullable<$Scalar.Int>
        float: $.Input.Nullable<$Scalar.Float>
        boolean: $.Input.Nullable<$Scalar.Boolean>
        id: $.Input.Nullable<$Scalar.ID>
      }>
    >
    fooBarUnion: $.Field<$.Output.Nullable<Union.FooBarUnion>, null>
    objectList: $.Field<$.Output.Nullable<$.Output.List<$.Output.Nullable<Object.Object1>>>, null>
    objectListNonNull: $.Field<$.Output.List<Object.Object1>, null>
    /**
     * Query enum field documentation.
     */
    abcEnum: $.Field<$.Output.Nullable<Enum.ABCEnum>, null>
    lowerCaseUnion: $.Field<$.Output.Nullable<Union.lowerCaseUnion>, null>
    interface: $.Field<$.Output.Nullable<Interface.Interface>, null>
    interfaceWithArgs: $.Field<
      $.Output.Nullable<Interface.Interface>,
      $.Args<{
        id: $Scalar.ID
      }>
    >
    interfaceNonNull: $.Field<Interface.Interface, null>
    unionFooBar: $.Field<$.Output.Nullable<Union.FooBarUnion>, null>
    unionFooBarWithArgs: $.Field<
      $.Output.Nullable<Union.FooBarUnion>,
      $.Args<{
        id: $.Input.Nullable<$Scalar.ID>
      }>
    >
    unionObject: $.Field<$.Output.Nullable<Object.ObjectUnion>, null>
    unionFooBarNonNull: $.Field<Union.FooBarUnion, null>
    unionObjectNonNull: $.Field<Object.ObjectUnion, null>
  }>
}

// ------------------------------------------------------------ //
//                             Enum                             //
// ------------------------------------------------------------ //

export namespace Enum {
  /**
   * Enum documentation.
   *
   * Members
   * "A" - (DEPRECATED: Enum value A is deprecated.)
   * "B" - Enum B member documentation.
   * "C" - Enum C member documentation. (DEPRECATED: Enum value C is deprecated.)
   */
  export type ABCEnum = $.Enum<'ABCEnum', ['A', 'B', 'C']>
}

// ------------------------------------------------------------ //
//                         InputObject                          //
// ------------------------------------------------------------ //

export namespace InputObject {
  export type InputObjectNested = $.InputObject<'InputObjectNested', {
    InputObject: $.Input.Nullable<InputObject.InputObject>
  }>

  export type InputObjectNestedNonNull = $.InputObject<'InputObjectNestedNonNull', {
    InputObject: InputObject.InputObject
  }>

  export type InputObject = $.InputObject<'InputObject', {
    id: $.Input.Nullable<$Scalar.ID>
    idRequired: $Scalar.ID
    date: $.Input.Nullable<$Scalar.Date>
    dateRequired: $Scalar.Date
  }>
}

// ------------------------------------------------------------ //
//                          Interface                           //
// ------------------------------------------------------------ //

export namespace Interface {
  export type DateInterface1 = $.Interface<'DateInterface1', {
    date1: $.Field<$.Output.Nullable<$Scalar.Date>, null>
  }, [Object.DateObject1]>

  export type Interface = $.Interface<'Interface', {
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
  }, [Object.Object1ImplementingInterface, Object.Object2ImplementingInterface]>
}

// ------------------------------------------------------------ //
//                            Object                            //
// ------------------------------------------------------------ //

export namespace Object {
  export type DateObject1 = $.Object$2<'DateObject1', {
    date1: $.Field<$.Output.Nullable<$Scalar.Date>, null>
  }>

  export type DateObject2 = $.Object$2<'DateObject2', {
    date2: $.Field<$.Output.Nullable<$Scalar.Date>, null>
  }>

  export type ObjectUnion = $.Object$2<'ObjectUnion', {
    fooBarUnion: $.Field<$.Output.Nullable<Union.FooBarUnion>, null>
  }>

  /**
   * Object documentation.
   */
  export type Foo = $.Object$2<'Foo', {
    /**
     * Field documentation.
     *
     * @deprecated Field a is deprecated.
     */
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
  }>

  export type Bar = $.Object$2<'Bar', {
    int: $.Field<$.Output.Nullable<$Scalar.Int>, null>
  }>

  export type ObjectNested = $.Object$2<'ObjectNested', {
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
    object: $.Field<$.Output.Nullable<Object.Object1>, null>
  }>

  export type lowerCaseObject = $.Object$2<'lowerCaseObject', {
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
  }>

  export type lowerCaseObject2 = $.Object$2<'lowerCaseObject2', {
    int: $.Field<$.Output.Nullable<$Scalar.Int>, null>
  }>

  export type Object1 = $.Object$2<'Object1', {
    string: $.Field<$.Output.Nullable<$Scalar.String>, null>
    int: $.Field<$.Output.Nullable<$Scalar.Int>, null>
    float: $.Field<$.Output.Nullable<$Scalar.Float>, null>
    boolean: $.Field<$.Output.Nullable<$Scalar.Boolean>, null>
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
  }>

  export type Object1ImplementingInterface = $.Object$2<'Object1ImplementingInterface', {
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
    int: $.Field<$.Output.Nullable<$Scalar.Int>, null>
  }>

  export type Object2ImplementingInterface = $.Object$2<'Object2ImplementingInterface', {
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
    boolean: $.Field<$.Output.Nullable<$Scalar.Boolean>, null>
  }>
}

// ------------------------------------------------------------ //
//                            Union                             //
// ------------------------------------------------------------ //

export namespace Union {
  export type DateUnion = $.Union<'DateUnion', [Object.DateObject1, Object.DateObject2]>

  /**
   * Union documentation.
   */
  export type FooBarUnion = $.Union<'FooBarUnion', [Object.Foo, Object.Bar]>

  export type lowerCaseUnion = $.Union<'lowerCaseUnion', [Object.lowerCaseObject, Object.lowerCaseObject2]>
}
"
`;

exports[`generates types from GraphQL SDL file 2`] = `
"import type * as CustomScalar from '../../../../_/customScalarCodecs.js'

declare global {
  interface SchemaCustomScalars {
    Date: CustomScalar.Date
  }
}

export * from '../../../../../src/Schema/Hybrid/types/Scalar/Scalar.js'
export * from '../../../../_/customScalarCodecs.js'
"
`;

exports[`generates types from GraphQL SDL file 3`] = `
"/* eslint-disable */

import * as $ from '../../../../../src/Schema/__.js'
import * as $Scalar from './Scalar.js'

export const ABCEnum = $.Enum(\`ABCEnum\`, [\`A\`, \`B\`, \`C\`])

export const InputObjectNested = $.InputObject(\`InputObjectNested\`, {
  InputObject: $.Input.field(() => $.Input.Nullable(InputObject)),
})

export const InputObjectNestedNonNull = $.InputObject(\`InputObjectNestedNonNull\`, {
  InputObject: $.Input.field(() => InputObject),
})

export const InputObject = $.InputObject(\`InputObject\`, {
  id: $.Input.field($.Input.Nullable($Scalar.ID)),
  idRequired: $.Input.field($Scalar.ID),
  date: $.Input.field($.Input.Nullable($Scalar.Date)),
  dateRequired: $.Input.field($Scalar.Date),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const DateObject1 = $.Object$(\`DateObject1\`, {
  date1: $.field($.Output.Nullable($Scalar.Date)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const DateObject2 = $.Object$(\`DateObject2\`, {
  date2: $.field($.Output.Nullable($Scalar.Date)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const ObjectUnion = $.Object$(\`ObjectUnion\`, {
  fooBarUnion: $.field($.Output.Nullable(() => FooBarUnion)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Foo = $.Object$(\`Foo\`, {
  id: $.field($.Output.Nullable($Scalar.ID)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Bar = $.Object$(\`Bar\`, {
  int: $.field($.Output.Nullable($Scalar.Int)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const ObjectNested = $.Object$(\`ObjectNested\`, {
  id: $.field($.Output.Nullable($Scalar.ID)),
  object: $.field($.Output.Nullable(() => Object1)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const lowerCaseObject = $.Object$(\`lowerCaseObject\`, {
  id: $.field($.Output.Nullable($Scalar.ID)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const lowerCaseObject2 = $.Object$(\`lowerCaseObject2\`, {
  int: $.field($.Output.Nullable($Scalar.Int)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Object1 = $.Object$(\`Object1\`, {
  string: $.field($.Output.Nullable($Scalar.String)),
  int: $.field($.Output.Nullable($Scalar.Int)),
  float: $.field($.Output.Nullable($Scalar.Float)),
  boolean: $.field($.Output.Nullable($Scalar.Boolean)),
  id: $.field($.Output.Nullable($Scalar.ID)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Object1ImplementingInterface = $.Object$(\`Object1ImplementingInterface\`, {
  id: $.field($.Output.Nullable($Scalar.ID)),
  int: $.field($.Output.Nullable($Scalar.Int)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Object2ImplementingInterface = $.Object$(\`Object2ImplementingInterface\`, {
  id: $.field($.Output.Nullable($Scalar.ID)),
  boolean: $.field($.Output.Nullable($Scalar.Boolean)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const DateUnion = $.Union(\`DateUnion\`, [DateObject1, DateObject2])

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const FooBarUnion = $.Union(\`FooBarUnion\`, [Foo, Bar])

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const lowerCaseUnion = $.Union(\`lowerCaseUnion\`, [lowerCaseObject, lowerCaseObject2])

export const DateInterface1 = $.Interface(\`DateInterface1\`, { date1: $.field($.Output.Nullable($Scalar.Date)) }, [
  DateObject1,
])
export const Interface = $.Interface(\`Interface\`, { id: $.field($.Output.Nullable($Scalar.ID)) }, [
  Object1ImplementingInterface,
  Object2ImplementingInterface,
])

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Query = $.Object$(\`Query\`, {
  date: $.field($.Output.Nullable($Scalar.Date)),
  dateNonNull: $.field($Scalar.Date),
  dateList: $.field($.Output.Nullable($.Output.List($.Output.Nullable($Scalar.Date)))),
  dateObject1: $.field($.Output.Nullable(() => DateObject1)),
  dateUnion: $.field($.Output.Nullable(() => DateUnion)),
  dateInterface1: $.field($.Output.Nullable(() => DateInterface1)),
  dateListNonNull: $.field($.Output.List($Scalar.Date)),
  dateArg: $.field($.Output.Nullable($Scalar.Date), $.Args({ date: $.Input.Nullable($Scalar.Date) })),
  dateArgNonNull: $.field($.Output.Nullable($Scalar.Date), $.Args({ date: $Scalar.Date })),
  dateArgList: $.field(
    $.Output.Nullable($Scalar.Date),
    $.Args({ date: $.Input.Nullable($.Input.List($.Input.Nullable($Scalar.Date))) }),
  ),
  dateArgNonNullList: $.field(
    $.Output.Nullable($Scalar.Date),
    $.Args({ date: $.Input.List($.Input.Nullable($Scalar.Date)) }),
  ),
  dateArgNonNullListNonNull: $.field($.Output.Nullable($Scalar.Date), $.Args({ date: $.Input.List($Scalar.Date) })),
  dateArgInputObject: $.field($.Output.Nullable($Scalar.Date), $.Args({ input: $.Input.Nullable(InputObject) })),
  InputObjectNested: $.field($.Output.Nullable($Scalar.ID), $.Args({ input: $.Input.Nullable(InputObjectNested) })),
  InputObjectNestedNonNull: $.field($.Output.Nullable($Scalar.ID), $.Args({ input: InputObjectNested })),
  id: $.field($.Output.Nullable($Scalar.ID)),
  idNonNull: $.field($Scalar.ID),
  string: $.field($.Output.Nullable($Scalar.String)),
  stringWithRequiredArg: $.field($.Output.Nullable($Scalar.String), $.Args({ string: $Scalar.String })),
  stringWithArgs: $.field(
    $.Output.Nullable($Scalar.String),
    $.Args({
      string: $.Input.Nullable($Scalar.String),
      int: $.Input.Nullable($Scalar.Int),
      float: $.Input.Nullable($Scalar.Float),
      boolean: $.Input.Nullable($Scalar.Boolean),
      id: $.Input.Nullable($Scalar.ID),
    }),
  ),
  stringWithArgEnum: $.field($.Output.Nullable($Scalar.String), $.Args({ ABCEnum: $.Input.Nullable(ABCEnum) })),
  stringWithListArg: $.field(
    $.Output.Nullable($Scalar.String),
    $.Args({ ints: $.Input.Nullable($.Input.List($.Input.Nullable($Scalar.Int))) }),
  ),
  stringWithListArgRequired: $.field(
    $.Output.Nullable($Scalar.String),
    $.Args({ ints: $.Input.List($.Input.Nullable($Scalar.Int)) }),
  ),
  stringWithArgInputObject: $.field(
    $.Output.Nullable($Scalar.String),
    $.Args({ input: $.Input.Nullable(InputObject) }),
  ),
  stringWithArgInputObjectRequired: $.field($.Output.Nullable($Scalar.String), $.Args({ input: InputObject })),
  listListIntNonNull: $.field($.Output.List($.Output.List($Scalar.Int))),
  listListInt: $.field(
    $.Output.Nullable($.Output.List($.Output.Nullable($.Output.List($.Output.Nullable($Scalar.Int))))),
  ),
  listInt: $.field($.Output.Nullable($.Output.List($.Output.Nullable($Scalar.Int)))),
  listIntNonNull: $.field($.Output.List($Scalar.Int)),
  object: $.field($.Output.Nullable(() => Object1)),
  objectNonNull: $.field(() => Object1),
  objectNested: $.field($.Output.Nullable(() => ObjectNested)),
  objectWithArgs: $.field(
    $.Output.Nullable(() => Object1),
    $.Args({
      string: $.Input.Nullable($Scalar.String),
      int: $.Input.Nullable($Scalar.Int),
      float: $.Input.Nullable($Scalar.Float),
      boolean: $.Input.Nullable($Scalar.Boolean),
      id: $.Input.Nullable($Scalar.ID),
    }),
  ),
  fooBarUnion: $.field($.Output.Nullable(() => FooBarUnion)),
  objectList: $.field($.Output.Nullable($.Output.List($.Output.Nullable(() => Object1)))),
  objectListNonNull: $.field($.Output.List(() => Object1)),
  abcEnum: $.field($.Output.Nullable(ABCEnum)),
  lowerCaseUnion: $.field($.Output.Nullable(() => lowerCaseUnion)),
  interface: $.field($.Output.Nullable(() => Interface)),
  interfaceWithArgs: $.field($.Output.Nullable(() => Interface), $.Args({ id: $Scalar.ID })),
  interfaceNonNull: $.field(() => Interface),
  unionFooBar: $.field($.Output.Nullable(() => FooBarUnion)),
  unionFooBarWithArgs: $.field($.Output.Nullable(() => FooBarUnion), $.Args({ id: $.Input.Nullable($Scalar.ID) })),
  unionObject: $.field($.Output.Nullable(() => ObjectUnion)),
  unionFooBarNonNull: $.field(() => FooBarUnion),
  unionObjectNonNull: $.field(() => ObjectUnion),
})

export const $Index = {
  Root: {
    Query,
    Mutation: null,
    Subscription: null,
  },
  objects: {
    DateObject1,
    DateObject2,
    ObjectUnion,
    Foo,
    Bar,
    ObjectNested,
    lowerCaseObject,
    lowerCaseObject2,
    Object1,
    Object1ImplementingInterface,
    Object2ImplementingInterface,
  },
  unions: {
    DateUnion,
    FooBarUnion,
    lowerCaseUnion,
  },
  interfaces: {
    DateInterface1,
    Interface,
  },
}
"
`;

exports[`schema2 1`] = `
"import type * as $ from '../../../../src/Schema/__.js'
import type * as $Scalar from './Scalar.ts'

// ------------------------------------------------------------ //
//                             Root                             //
// ------------------------------------------------------------ //

export namespace Root {
  export type Mutation = $.Object$2<'Mutation', {
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
    idNonNull: $.Field<$Scalar.ID, null>
  }>

  export type Query = $.Object$2<'Query', {
    date: $.Field<$.Output.Nullable<$Scalar.Date>, null>
    dateArg: $.Field<
      $.Output.Nullable<$Scalar.Date>,
      $.Args<{
        date: $.Input.Nullable<$Scalar.Date>
      }>
    >
    dateArgNonNull: $.Field<
      $.Output.Nullable<$Scalar.Date>,
      $.Args<{
        date: $Scalar.Date
      }>
    >
    dateNonNull: $.Field<$Scalar.Date, null>
    dateObject1: $.Field<$.Output.Nullable<Object.DateObject1>, null>
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
    idNonNull: $.Field<$Scalar.ID, null>
    interface: $.Field<$.Output.Nullable<Interface.Interface>, null>
    objectWithArgs: $.Field<
      $.Output.Nullable<Object.Object1>,
      $.Args<{
        boolean: $.Input.Nullable<$Scalar.Boolean>
        float: $.Input.Nullable<$Scalar.Float>
        id: $.Input.Nullable<$Scalar.ID>
        int: $.Input.Nullable<$Scalar.Int>
        string: $.Input.Nullable<$Scalar.String>
      }>
    >
    unionFooBar: $.Field<$.Output.Nullable<Union.FooBarUnion>, null>
  }>
}

// ------------------------------------------------------------ //
//                             Enum                             //
// ------------------------------------------------------------ //

export namespace Enum {
  // -- no types --
}

// ------------------------------------------------------------ //
//                         InputObject                          //
// ------------------------------------------------------------ //

export namespace InputObject {
  // -- no types --
}

// ------------------------------------------------------------ //
//                          Interface                           //
// ------------------------------------------------------------ //

export namespace Interface {
  export type Interface = $.Interface<'Interface', {
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
  }, [Object.Object1ImplementingInterface, Object.Object2ImplementingInterface]>
}

// ------------------------------------------------------------ //
//                            Object                            //
// ------------------------------------------------------------ //

export namespace Object {
  export type Bar = $.Object$2<'Bar', {
    int: $.Field<$.Output.Nullable<$Scalar.Int>, null>
  }>

  export type DateObject1 = $.Object$2<'DateObject1', {
    date1: $.Field<$.Output.Nullable<$Scalar.Date>, null>
  }>

  /**
   * Object documentation.
   */
  export type Foo = $.Object$2<'Foo', {
    /**
     * Field documentation.
     *
     * @deprecated Field a is deprecated.
     */
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
  }>

  export type Object1 = $.Object$2<'Object1', {
    boolean: $.Field<$.Output.Nullable<$Scalar.Boolean>, null>
    float: $.Field<$.Output.Nullable<$Scalar.Float>, null>
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
    int: $.Field<$.Output.Nullable<$Scalar.Int>, null>
    string: $.Field<$.Output.Nullable<$Scalar.String>, null>
  }>

  export type Object1ImplementingInterface = $.Object$2<'Object1ImplementingInterface', {
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
    int: $.Field<$.Output.Nullable<$Scalar.Int>, null>
  }>

  export type Object2ImplementingInterface = $.Object$2<'Object2ImplementingInterface', {
    boolean: $.Field<$.Output.Nullable<$Scalar.Boolean>, null>
    id: $.Field<$.Output.Nullable<$Scalar.ID>, null>
  }>
}

// ------------------------------------------------------------ //
//                            Union                             //
// ------------------------------------------------------------ //

export namespace Union {
  /**
   * Union documentation.
   */
  export type FooBarUnion = $.Union<'FooBarUnion', [Object.Bar, Object.Foo]>
}
"
`;

exports[`schema2 2`] = `
"import type * as CustomScalar from '../../customScalarCodecs.js'

declare global {
  interface SchemaCustomScalars {
    Date: CustomScalar.Date
  }
}

export * from '../../../../src/Schema/Hybrid/types/Scalar/Scalar.js'
export * from '../../customScalarCodecs.js'
"
`;

exports[`schema2 3`] = `
"/* eslint-disable */

import * as $ from '../../../../src/Schema/__.js'
import * as $Scalar from './Scalar.js'

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Bar = $.Object$(\`Bar\`, {
  int: $.field($.Output.Nullable($Scalar.Int)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const DateObject1 = $.Object$(\`DateObject1\`, {
  date1: $.field($.Output.Nullable($Scalar.Date)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Foo = $.Object$(\`Foo\`, {
  id: $.field($.Output.Nullable($Scalar.ID)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Object1 = $.Object$(\`Object1\`, {
  boolean: $.field($.Output.Nullable($Scalar.Boolean)),
  float: $.field($.Output.Nullable($Scalar.Float)),
  id: $.field($.Output.Nullable($Scalar.ID)),
  int: $.field($.Output.Nullable($Scalar.Int)),
  string: $.field($.Output.Nullable($Scalar.String)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Object1ImplementingInterface = $.Object$(\`Object1ImplementingInterface\`, {
  id: $.field($.Output.Nullable($Scalar.ID)),
  int: $.field($.Output.Nullable($Scalar.Int)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Object2ImplementingInterface = $.Object$(\`Object2ImplementingInterface\`, {
  boolean: $.field($.Output.Nullable($Scalar.Boolean)),
  id: $.field($.Output.Nullable($Scalar.ID)),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const FooBarUnion = $.Union(\`FooBarUnion\`, [Bar, Foo])

export const Interface = $.Interface(\`Interface\`, { id: $.field($.Output.Nullable($Scalar.ID)) }, [
  Object1ImplementingInterface,
  Object2ImplementingInterface,
])

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Mutation = $.Object$(\`Mutation\`, {
  id: $.field($.Output.Nullable($Scalar.ID)),
  idNonNull: $.field($Scalar.ID),
})

// @ts-ignore - circular types cannot infer. Ignore in case there are any. This comment is always added, it does not indicate if this particular type could infer or not.
export const Query = $.Object$(\`Query\`, {
  date: $.field($.Output.Nullable($Scalar.Date)),
  dateArg: $.field($.Output.Nullable($Scalar.Date), $.Args({ date: $.Input.Nullable($Scalar.Date) })),
  dateArgNonNull: $.field($.Output.Nullable($Scalar.Date), $.Args({ date: $Scalar.Date })),
  dateNonNull: $.field($Scalar.Date),
  dateObject1: $.field($.Output.Nullable(() => DateObject1)),
  id: $.field($.Output.Nullable($Scalar.ID)),
  idNonNull: $.field($Scalar.ID),
  interface: $.field($.Output.Nullable(() => Interface)),
  objectWithArgs: $.field(
    $.Output.Nullable(() => Object1),
    $.Args({
      boolean: $.Input.Nullable($Scalar.Boolean),
      float: $.Input.Nullable($Scalar.Float),
      id: $.Input.Nullable($Scalar.ID),
      int: $.Input.Nullable($Scalar.Int),
      string: $.Input.Nullable($Scalar.String),
    }),
  ),
  unionFooBar: $.field($.Output.Nullable(() => FooBarUnion)),
})

export const $Index = {
  Root: {
    Query,
    Mutation,
    Subscription: null,
  },
  objects: {
    Bar,
    DateObject1,
    Foo,
    Object1,
    Object1ImplementingInterface,
    Object2ImplementingInterface,
  },
  unions: {
    FooBarUnion,
  },
  interfaces: {
    Interface,
  },
}
"
`;
